### PR TITLE
Casper/core 765/integrate lmdb block store

### PR DESCRIPTION
**Pull request by dzajkowski**
## Overview
Introduces LMDB BlockStore to replace InMemBS.
Current implementation does not carry state over.

### Which JIRA issue does this PR relate to? If there is not a JIRA issue addressing this work, please create one now and add the link here.
https://rchain.atlassian.net/browse/CORE-765

### Complete this checklist before you submit the PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### Notes
none

### CODE REVIEW

File: node/src/main/scala/coop/rchain/node/node.scala [hunk 1]
```diff
@@ -181,33 +180,35 @@ class NodeRuntime(conf: Conf)(implicit scheduler: Scheduler) {
   implicit val nodeDiscoveryEffect: NodeDiscovery[Task] =
     new KademliaNodeDiscovery[Task](src, defaultTimeout)
 
+  if (Files.notExists(casperPersistencePath)) Files.createDirectories(casperPersistencePath)
+  val config =
+    coop.rchain.blockstorage.LMDBBlockStore.Config(casperPersistencePath, 1024 * 1024 * 1024)
+  val syncEffect: Sync[Effect] = SyncInstances.syncEffect
+  implicit val blockStore: BlockStore[Effect] =
+    LMDBBlockStore.create[Effect](config)(syncEffect, metricsEffect)
```
Comment (dzajkowski) [inline node/src/main/scala/coop/rchain/node/node.scala]: not convinced this is a good spot for this init, but did not want to pollute the `Runtime`

File: block-storage/src/main/scala/coop/rchain/blockstorage/InMemBlockStore.scala [hunk 2]
```diff
@@ -42,11 +42,13 @@ class InMemBlockStore[F[_], E] private ()(implicit
           }
     } yield ()
 
-  private[blockstorage] def clear(): F[Unit] =
+  def clear(): F[Unit] =
     for {
       _ <- metricsF.incrementCounter("block-store-put")
```
Comment (lukasz-golebiewski) [inline block-storage/src/main/scala/coop/rchain/blockstorage/InMemBlockStore.scala]: This metric name doesn't look right

File: casper/src/test/scala/coop/rchain/casper/CliqueOracleTest.scala [hunk 3]
```diff
@@ -24,12 +24,11 @@ import monix.execution.Scheduler.Implicits.global
 
 import scala.collection.immutable.{HashMap, HashSet}
 
-class CliqueOracleTest extends FlatSpec with Matchers with BlockGenerator {
+class CliqueOracleTest extends FlatSpec with Matchers with BlockGenerator with WithBlockStore {
```
Comment (lukasz-golebiewski) [inline casper/src/test/scala/coop/rchain/casper/CliqueOracleTest.scala]: `with WithBlockStore` sound redundant. `with BlockStoreFixture` maybe?

File: node/src/main/scala/coop/rchain/node/node.scala [hunk 4]
```diff
@@ -181,33 +180,35 @@ class NodeRuntime(conf: Conf)(implicit scheduler: Scheduler) {
   implicit val nodeDiscoveryEffect: NodeDiscovery[Task] =
     new KademliaNodeDiscovery[Task](src, defaultTimeout)
 
+  if (Files.notExists(casperPersistencePath)) Files.createDirectories(casperPersistencePath)
```
Comment (birchmd) [inline node/src/main/scala/coop/rchain/node/node.scala]: I would move this directory existence check into the `LMDBBlockStore.create` method itself.

File: node/src/main/scala/coop/rchain/node/node.scala [hunk 5]
```diff
@@ -256,6 +257,8 @@ class NodeRuntime(conf: Conf)(implicit scheduler: Scheduler) {
     resources.runtime.close()
     println("Shutting down Casper runtime ...")
     resources.casperRuntime.close()
+    println("Bringing BlockStore down ...")
```
Comment (lukasz-golebiewski) [inline node/src/main/scala/coop/rchain/node/node.scala]: https://www.youtube.com/watch?v=KYh7PwDo3Iw

File: node/src/main/scala/coop/rchain/node/node.scala [hunk 6]
```diff
@@ -181,33 +180,35 @@ class NodeRuntime(conf: Conf)(implicit scheduler: Scheduler) {
   implicit val nodeDiscoveryEffect: NodeDiscovery[Task] =
     new KademliaNodeDiscovery[Task](src, defaultTimeout)
 
+  if (Files.notExists(casperPersistencePath)) Files.createDirectories(casperPersistencePath)
+  val config =
+    coop.rchain.blockstorage.LMDBBlockStore.Config(casperPersistencePath, 1024 * 1024 * 1024)
```
Comment (lukasz-golebiewski) [inline node/src/main/scala/coop/rchain/node/node.scala]: `1024 * 1024 * 1024` should be taken from the config

Comment (birchmd) [inline node/src/main/scala/coop/rchain/node/node.scala]: The store size should probably also be configurable. I chose to construct the Casper config object from the node config object in [`conf.scala`](https://github.com/rchain/rchain/blob/dev/node/src/main/scala/coop/rchain/node/conf.scala#L247). You could consider doing the same.

File: node/src/main/scala/coop/rchain/node/node.scala (hunk 1 above)
Comment (birchmd) [reply node/src/main/scala/coop/rchain/node/node.scala]: If you make the two changes above then this just becomes a one-liner similar to the other effect instantiations above, so then it should be fine.

File: casper/src/test/scala/coop/rchain/casper/helper/WithBlockStore.scala [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+          flags: List[EnvFlags] = List(EnvFlags.MDB_NOTLS)): Env[ByteBuffer] =
+    Env
+      .create()
+      .setMapSize(mapSize)
+      .setMaxDbs(8)
+      .setMaxReaders(126)
+      .open(path.toFile, flags: _*)
+
+  def dbDir: Path   = Files.createTempDirectory("casper-block-store-test-")
+  val mapSize: Long = 1024L * 1024L * 4096L
```
Comment (birchmd) [inline casper/src/test/scala/coop/rchain/casper/helper/WithBlockStore.scala]: I doubt the test block store will need to be this big! 100 MB is probably sufficient.

**Pull request merged**